File: app/src/main/java/module/RoomModule.java
package module;

import android.app.Application;

import androidx.room.Room;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import room.PatientDao;
import room.UserDao;
import room.UserRoomDatabase;

@Module
public class RoomModule {

    private UserRoomDatabase userRoomDatabase;

    public RoomModule(Application mApplication) {
        userRoomDatabase = Room.databaseBuilder(mApplication, UserRoomDatabase.class, "demo-db").build();
    }

    @Singleton
    @Provides
    UserRoomDatabase providesUserRoomDatabase() {
        return userRoomDatabase;
    }

    @Singleton
    @Provides
    UserDao providesUserDao(UserRoomDatabase userRoomDatabase) {
        return userRoomDatabase.userDao();
    }

    @Singleton
    @Provides
    PatientDao providesPatientDao(UserRoomDatabase userRoomDatabase) {
        return userRoomDatabase.patientDao();
    }

}

File: app/src/main/java/module/RoomModule.kt
package module

import android.app.Application
import androidx.room.Room
import dagger.Module
import dagger.Provides
import room.PatientDao
import room.UserDao
import room.UserRoomDatabase
import javax.inject.Singleton


@Module
class RoomModule(mApplication: Application) {
    private val userRoomDatabase: UserRoomDatabase

    @Singleton
    @Provides
    fun providesUserRoomDatabase(): UserRoomDatabase {
        return userRoomDatabase
    }

    @Singleton
    @Provides
    fun providesUserDao(userRoomDatabase: UserRoomDatabase): UserDao {
        return userRoomDatabase.userDao()
    }

    @Singleton
    @Provides
    fun providesPatientDao(userRoomDatabase: UserRoomDatabase): PatientDao {
        return userRoomDatabase.patientDao()
    }

    init {
        userRoomDatabase = Room.databaseBuilder(
            mApplication!!,
            UserRoomDatabase::class.java, "demo-db"
        ).build()
    }
}
